driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -135,13 +145,27 @@ public ServerDescription getInitialServerDescription() {
        return initialServerDescription;
     }
 
+    @Override
+    public int getGeneration() {
+        return generation;
+    }
+
     @Override
     public void open() {
         isTrue("Open already called", stream == null);
         stream = streamFactory.create(serverId.getAddress());
         try {
             stream.open();
-            InternalConnectionInitializationDescription initializationDescription = connectionInitializer.initialize(this);
+
+            InternalConnectionInitializationDescription initializationDescription = connectionInitializer.startHandshake(this);
```

Comment:
We split the handshake into two parts because of the requirement that we invalidate connections with the same serviceId if open fails once we know the serviceId.  So in the first part we just do the isMaster/hello call, which returns the serviceId.  Once we have it we set the generation from the generation supplier.  Then we complete the handshake.

It's a bit annoying...